### PR TITLE
Updates for new OpenAI release post dev day

### PR DESCRIPTION
**Pull request by transitive-bullshit**
- adds support for tools
- changes some other types required by openai updates

### CODE REVIEW

File: src/prompt/utils/message.ts [hunk 1]
```diff
@@ -98,21 +98,50 @@ export class Msg {
     return { role: 'function', content: contentString, name };
   }
 
+  /** Create a function call message with argumets. */
+  static toolCalls(
+    tool_calls: Prompt.Msg.ToolCall[],
+    opts?: {
+      /** The name descriptor for the message.(message.name) */
+      name?: string;
```
Comment (transitive-bullshit) [inline src/prompt/utils/message.ts:106]: it looks like openai may have removed support for `name`, i'm guessing in favor of the stateful api for that use case

File: src/model/chat.test.ts [hunk 2]
```diff
@@ -12,7 +12,7 @@ const FAKE_RESPONSE: Model.Chat.Response = {
   created: 0,
   id: 'fake-id',
   model: 'gpt-fake',
-  object: 'fake-chat',
+  object: 'chat.completion',
```
Comment (transitive-bullshit) [inline src/model/chat.test.ts:15]: this seems to be required now typing-wise

File: src/prompt/utils/message.ts (hunk 1 above)
Comment (rileytomasek) [reply src/prompt/utils/message.ts:106]: Yeah, it does look gone. I also noticed they started naming the different message type [here](https://platform.openai.com/docs/api-reference/chat/create). We should probably stay in line with those to avoid confusion, but I still prefer the even more explicit assistant breakdown. What do you think of:
- System
- User
- Assistant
  - AssistantTool
  - AssistantFunction
- Tool
- Function

**Review by rileytomasek — CHANGES_REQUESTED** [review]
I'm not sure this is the best way to think about the relationship between the legacy functions and the new tools.

It seems like the legacy functions have the same structure and function, but are now nested under tools. I think all of the schema/spec/parser/implementation parts of `AIFunction` should stay the same (without tools) and should be used by a new higher level tools abstraction that is just wraps a function for now, and presumably other things in the future.

Also, can you please start writing real commit messages :)

**Comment by transitive-bullshit** [comment]
> It seems like the legacy functions have the same structure and function, but are now nested under tools. I think all of the schema/spec/parser/implementation parts of AIFunction should stay the same (without tools) and should be used by a new higher level tools abstraction that is just wraps a function for now, and presumably other things in the future.

Sounds good & agreed. I haven't actually used `createAIFunction` yet, so that would be a good first step haha

> Also, can you please start writing real commit messages :)

Yep; sorry about that. I try to use real git-flow commit messages for substantive stuff and `commit-emoji` for minor things. Will work on breaking that habit for dexa stuff. :)

File: src/prompt/utils/message.ts (hunk 1 above)
Comment (transitive-bullshit) [reply src/prompt/utils/message.ts:106]: Sounds good; just including a screenshot to make sure we're talking about the same naming.

I'm fine w/ the more explicit version. I was just following the pattern you had done for functions previously as best I could pattern-match.

CleanShot 2023-11-07 at 15 41 31@2x

Comment (transitive-bullshit) [reply src/prompt/utils/message.ts:106]: Also dealing w/ this which seems to be a typing bug in the latest `openai` release https://github.com/openai/openai-node/issues/462

**Comment by transitive-bullshit** [comment]
This PR is now dependent on https://github.com/dexaai/openai-fetch/pull/38

**Comment by transitive-bullshit** [comment]
@rileytomasek I changed the tools and function `assistant` types and utility functions to have an explicit `assistant` prefix.

I also fixed all the lingering TS issues, mainly due to the `message` stuff. Such a PITA.

Not 100% sure I incorporated the `assistant` prefix feedback you gave as you were expecting. Happy to adjust.

Ahhh just remembered I need to fix the `AIFunction` to not have tools attached..

File: src/model/chat.ts [hunk 3]
```diff
@@ -131,12 +131,13 @@ export class ChatModel extends AbstractModel<
...
       const response: Model.Chat.ApiResponse = {
         ...chunk,
+        object: 'chat.completion',
         choices: [
           {
             finish_reason:
               choice.finish_reason as Model.Chat.Response['choices'][0]['finish_reason'],
             index: choice.index,
-            message: choice.delta as Model.Message,
+            message: choice.delta as Model.Message & { role: 'assistant' },
```
Comment (rileytomasek) [inline src/model/chat.ts:140]: Do we need this now that you fixed the types for `openai-fetch`?

File: src/model/clients/openai.ts
Comment (rileytomasek) [inline src/model/clients/openai.ts:1]: Why this change? We are using this internally and I'd rather not change the existing code unless there is a reason for this.

**Review by rileytomasek — COMMENTED** [review]
This looks good other than the name changes. Can we hold off removing the `Message.name` stuff until everything settles from the new releases and it's clear what the status of name is? We're using it internally and I'd rather not change until we know it's definitely gone.

File: src/model/chat.ts (hunk 3 above)
Comment (transitive-bullshit) [reply src/model/chat.ts:140]: Good question && yes, unfortunately TS complains without it.

One thing I've noticed is that the new `openai` types seem to have gotten a lot stricter in terms of inputs/outputs used in different places. I could be wrong since I've only just started getting back into things.

Comment (transitive-bullshit) [reply src/model/chat.ts:140]: CleanShot 2023-11-08 at 14 45 45@2x

File: src/model/clients/openai.ts
Comment (transitive-bullshit) [reply src/model/clients/openai.ts:1]: hmmm was just a part of my removing `name`, and I thought it made more sense to have a util to format the `name` identifier directly than an object which at the time I was thinking would no longer have a `name` property.

all good && will revert

**Comment by transitive-bullshit** [comment]
> This looks good other than the name changes. Can we hold off removing the `Message.name` stuff until everything settles from the new releases and it's clear what the status of name is? We're using it internally and I'd rather not change until we know it's definitely gone.

Done.

> Edit: these changes are also going to require a major version bump, right?

Yes, though it's a brand new package, so we could just do a minor? I'm fine either way.

**Comment by transitive-bullshit** [comment]
Fixes https://github.com/dexaai/dexter/issues/8 as well

File: src/model/chat.ts (hunk 3 above)
Comment (rileytomasek) [reply src/model/chat.ts:140]: i think this is just from their jank types so we can leave as is for now

File: src/prompt/types.ts [hunk 4]
```diff
@@ -73,40 +81,60 @@ export namespace Prompt {
...
+      /** The type of the tool. Currently, only `function` is supported. */
+      type: 'function';
+      /** The function that the model called. */
+      function: FunctionCall;
+    };
+
     /** Message with text content for the system. */
     export type System = {
       role: 'system';
-      name?: string;
```
Comment (rileytomasek) [inline src/prompt/types.ts:79]: You missed these names when reverting

**Comment by rileytomasek** [comment]
Everything looks good, except for the names I suggested 😂

What do you think of:
```
System
User
Assistant
FuncCall
FuncResult
ToolCall
ToolResult
```

**Comment by transitive-bullshit** [comment]
> Everything looks good, except for the names I suggested 😂
> 
> What do you think of:
> 
> ```
> System
> User
> Assistant
> FuncCall
> FuncResult
> ToolCall
> ToolResult
> ```
> 
> And then mirror to `Msg` methods like:
> 
> ```
> Msg.system()
> Msg.user()
> Msg.assistant()
> Msg.funcCall()
> Msg.funcResult()
> Msg.toolCall()
> Msg.toolResult()
> 
> Msg.isSystem()
> Msg.isUser()
> Msg.isAssistant()
> Msg.isFuncCall()
> Msg.isFuncResult()
> Msg.isToolCall()
> Msg.isToolResult()
> ```

I'm not opinionated when it comes to naming; the only problem I see w/ this is that `ToolCall` type represents both the individual tool call `{ id: string, type: string, function: FunctionCall }` as well as the tool call message type, which represents an array of tool calls. 

It's the same issue w/ `FunctionCall` type except you got around that by shortening the message function call type to `FuncCall` to avoid the naming collision.

I don't have a strong feeling either way.

**Comment by transitive-bullshit** [comment]
If we call the tool calls message `ToolCalls` or `ToolsCall`, that would get around it. `ToolsCall` sounds better, but it's weird for the underlying field to be `tool_calls` and then refer to it as a `ToolsCall`...

File: src/prompt/types.ts (hunk 4 above)
Comment (transitive-bullshit) [reply src/prompt/types.ts:79]: ahhh good catch :) will revert

**Review by rileytomasek — APPROVED** [review]
🚢
